subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/processing/AnnotationProcessorDetector.java

```diff
@@ -60,10 +61,15 @@ public AnnotationProcessorDetector(FileContentCacheFactory cacheFactory, Logger
         this.logStackTraces = logStackTraces;
     }
 
-    public List<AnnotationProcessorDeclaration> detectProcessors(Iterable<File> processorPath) {
-        List<AnnotationProcessorDeclaration> processors = Lists.newArrayList();
+    public Map<String, AnnotationProcessorDeclaration> detectProcessors(Iterable<File> processorPath) {
+        Map<String, AnnotationProcessorDeclaration> processors = Maps.newLinkedHashMap();
         for (File jarOrClassesDir : processorPath) {
-            processors.addAll(cache.get(jarOrClassesDir));
+            for (AnnotationProcessorDeclaration declaration : cache.get(jarOrClassesDir)) {
+                String className = declaration.getClassName();
+                if (!processors.containsKey(className)) {
```

Comment:
:o: What if it already contains the key? It seems there's a hidden assumption here. Either you can directly store, because you assume that for a single class name, it's always going to be the same processor type (sounds reasonable, unless 2 versions of the same processor are on path). Or, you assume that they could be different, in which case at least a warning should be issued (duplicate processor, ...).

Reply:
This mirrors what `ServiceLoader` would do, i.e. the way javac would load processors. If there is a second one later on the classpath it just silently ignores it, so I think we should do the same.